chapter_natural-language-processing-pretraining/word-embedding-dataset.ipynb

```diff
@@ -0,0 +1,825 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "Integer[][] tinyDataset =\n",
+    "        new Integer[][] {\n",
+    "            IntStream.range(0, 7)\n",
+    "                    .boxed()\n",
+    "                    .collect(Collectors.toList())\n",
+    "                    .toArray(new Integer[] {}),\n",
+    "            IntStream.range(7, 10)\n",
+    "                    .boxed()\n",
+    "                    .collect(Collectors.toList())\n",
+    "                    .toArray(new Integer[] {})\n",
+    "        };\n",
+    "\n",
+    "System.out.println(\"dataset \" + Arrays.deepToString(tinyDataset));\n",
+    "Pair<ArrayList<Integer>, ArrayList<ArrayList<Integer>>> centerContextPair =\n",
+    "        getCentersAndContext(tinyDataset, 2);\n",
+    "for (int i = 0; i < centerContextPair.getValue().size(); i++) {\n",
+    "    System.out.println(\n",
+    "            \"Center \"\n",
+    "                    + centerContextPair.getKey().get(i)\n",
+    "                    + \" has contexts\"\n",
+    "                    + centerContextPair.getValue().get(i));\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 21
+   },
+   "source": [
+    "We set the maximum context window size to 5. The following extracts all the central target words and their context words in the dataset.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "centerContextPair = getCentersAndContext(corpus, 5);\n",
+    "ArrayList<Integer> allCenters = centerContextPair.getKey();\n",
+    "ArrayList<ArrayList<Integer>> allContexts = centerContextPair.getValue();\n",
+    "System.out.println(\"# center-context pairs:\" + allCenters.size());"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 23
+   },
+   "source": [
+    "### Negative Sampling\n",
+    "\n",
+    "We use negative sampling for approximate training. For a central and context word pair, we randomly sample $K$ noise words ($K=5$ in the experiment). According to the suggestion in the Word2vec paper, the noise word sampling probability $P(w)$ is the ratio of the word frequency of $w$ to the total word frequency raised to the power of 0.75 :cite:`Mikolov.Sutskever.Chen.ea.2013`.\n",
+    "\n",
+    "We first define a class to draw a candidate according to the sampling weights. It caches a 10,000 size random number bank.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public class RandomGenerator {\n",
+    "    /* Draw a random int in [0, n] according to n sampling weights. */\n",
+    "\n",
+    "    private List<Integer> population;\n",
+    "    private List<Double> samplingWeights;\n",
+    "    private List<Integer> candidates;\n",
+    "    private List<org.apache.commons.math3.util.Pair<Integer, Double>> pmf;\n",
```

Comment:
use djl pair

Reply:
I am using that specific pair because it works interchangeably with EnumeratedDistribution